### ISSUE TITLE
termux-api commands hang in terminal without any output

### ISSUE THREAD

**Issue by johnwuelk** [termux/termux-api#466]
**Problem description**

termux-api commands do not work and hang in terminal without any visible output. The only thing I can do is Control+c for cancelling the command.
According to forum posts, I'm not the only encountering this issue.

**Steps to reproduce**

Newest version of termux-api and termux is installed from F-Droid and all permissions are granted in the app settings for both apps. Plus, battery optimization is turned off for all termux apps, so that termux apps are granted to run unlimited at all times. termux-battery-status is weirdly only working sometimes, it has only worked 2 or 3 times out of the many I tried. The other commands do not work at all. Rebooting and/or reinstalling the apps did not lead to any solution. 

**Expected behavior**

Regular output of termux-api specific commands like termux-battery-status

**Additional information**
The phones were NOT rooted. Does the Android device really have to be rooted for termux-api to (fully) work all the time?

* Termux application version: 0.117 and termux-api version 0.56
* Android OS version: 9 and 10
* Device model: Blu G60 and Blu G90

**Comment by Grimler91** [comment]
Do you have the termux-api app installed? You mention the installed version of the termux-api package, but not of the app

**Comment by johnwuelk** [comment]
Yes, I have the termux-api APP from F-Droid installed as well, not only the package within termux.

It does work without root on other phones like Samsung but sadly not on my Blu phone.

**Comment by Grimler91** [comment]
Do you get any errors/warnings/related output in `adb logcat` when trying to use it?

**Comment by johnwuelk** [comment]
Yes:
```
~ $ adb logcat
```
* daemon not running; starting now at tcp:5037
* daemon started successfully
- waiting for device -

**Comment by agnostic-apollo** [comment]
> Yes:

Generate a full dump after connecting phone to pc, running api command and then running command on pc: `adb shell -d > logcat.txt`

https://developer.android.com/studio/command-line/logcat

You should also disable battery optimizations and duraspeed, etc.

https://dontkillmyapp.com/

**Comment by johnwuelk** [comment]
`adb shell logcat -d > logcat.txt` worked and generated this following file:
[logcat.txt](https://github.com/termux/termux-api/files/7825531/logcat.txt)
`termux-battery-status` was running on phone before I ran above command.

if this is relevant the output of `termux-api-start` was:
Starting service: Intent { cmp=com.termux.api/.KeepAliveService }

**Comment by john-peterson** [comment]
```
termux-api-start
```
I did not even have a clue about this command

And btw can you extend the expiration of action files? The latest and all previous have expired

**Comment by agnostic-apollo** [comment]
>The latest and all previous have expired

https://github.com/termux/termux-api/actions/runs/5930893260

**Comment by rocket-pig** [comment]
Hi, I appreciate response - I have both installed and actually reinstalled both before posting just to double make sure. Still same result. I can try the -f hook if its helpful? (Is that what youre saying, or that it also wouldnt be?)

**Comment by rocket-pig** [comment]
its 1.5MB of text, pastebin has no upload function and other sites are giving me grief as well. It ends with many many copies of the mprotect line...
Im not knowlegable enough to decipher what this means or if its helpful in troubleshooting why this is happening for folks. 

```
[pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ) = 0         [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ|PROT_WRITE) = 0                                                              [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ) = 0         [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ|PROT_WRITE) = 0                                                              [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ) = 0         [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ|PROT_WRITE) = 0                                                              [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ) = 0         [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ|PROT_WRITE) = 0                                                              [pid 13823] mprotect(0x79c43cd000, 4096, PROT_READ) = 0         [pid 13823] munmap(0x79c313e000, 4096)  = 0                     [pid 13823] munmap(0x79c3271000, 4096)  = 0                     [pid 13823] munmap(0x79c3274000, 4096)  = 0                     [pid 13823] munmap(0x79c43cc000, 4096)  = 0                     [pid 13823] munmap(0x79c43cd000, 4096)  = 0                     [pid 13823] exit_group(0 <unfinished ...>                       [pid 13834] <... futex resumed>)        = ? <unavailable>       [pid 13840] <... ioctl resumed>)        = ? <unavailable>       [pid 13833] <... futex resumed>)        = ?                     [pid 13832] <... futex resumed>)        = ?                     [pid 13830] <... futex resumed>)        = ? <unavailable>       [pid 13823] <... exit_group resumed>)   = ?                     [pid 13841] <... ioctl resumed> <unfinished ...>) = ?           [pid 13839] <... futex resumed>)        = ?                     [pid 13838] <... futex resumed>)        = ?                     [pid 13836] <... futex resumed>)        = ?                     [pid 13835] <... rt_sigtimedwait resumed> <unfinished ...>) = ? [pid 13837] <... futex resumed>)        = ?                     [pid 13833] +++ exited with 0 +++                               [pid 13832] +++ exited with 0 +++                               [pid 13831] <... futex resumed>)        = ?                     [pid 13841] +++ exited with 0 +++                               [pid 13840] +++ exited with 0 +++                               [pid 13839] +++ exited with 0 +++                               [pid 13838] +++ exited with 0 +++                               [pid 13837] +++ exited with 0 +++                               [pid 13836] +++ exited with 0 +++                               [pid 13835] +++ exited with 0 +++                               [pid 13834] +++ exited with 0 +++                               [pid 13831] +++ exited with 0 +++                               [pid 13830] +++ exited with 0 +++                               [pid 13823] +++ exited with 0 +++                               [pid 13820] <... accept4 resumed>0x7fd59b89f0, [110], 0) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
```

**Comment by agnostic-apollo** [comment]
Long hold on terminal and then `Report Issue` with debug enabled. Then save file from options button in report activity.

**Comment by rocket-pig** [comment]
> Long hold on terminal and then `Report Issue` with debug enabled. Then save file from options button in report activity.

...heres the result of that (400K+):

https://pastebin.com/awxp0Xbh

**Comment by rocket-pig** [comment]
...and just to skip that possible suggestion, i see it says in there that termux-am package is updateable, so i did, and ran termux-api stop; termux-api start; termux-sms-list and...its been over a minute with no response at all still.

**Comment by rocket-pig** [comment]
> @rocket-pig To catch the long strace log, use `strace -f -o strace-log.txt your-command`. Then upload file `strace-log.txt`.
> 
> `mprotect` lines possibly related to `/system/bin/app_process`. This utility will generate lots of calls.

Tldr my post above, i did that and its multiple MBs and couldn't get it here or on any paste sites.

**Comment by agnostic-apollo** [comment]
Change log level to `Verbose` for Termux:API and then run api command and then run `adb logcat -d > logcat.txt`. **After that**, install Termux:API app from [here](https://github.com/termux/termux-api/actions/runs/5947876837) and then start app from launcher and then try running command again.

https://github.com/termux/termux-app#debugging

**Comment by rocket-pig** [comment]
I appreciate the help, updating to that version fixed the problem, maybe could have led with that lol.

FWIW Chrome seems incapable of clicking the .apk links on those Action pages, but in Brave browser they always work. 

Thanks again.

**Comment by agnostic-apollo** [comment]
Welcome. Logs would have been useful to find real issue.

Probably aren't logged into github on chrome.

**Comment by machitgarha** [comment]
> Change log level to `Verbose` for Termux:API and then run api command and then run `adb logcat -d > logcat.txt`. **After that**, install Termux:API app from [here](https://github.com/termux/termux-api/actions/runs/5947876837) and then start app from launcher and then try running command again.
> 
> https://github.com/termux/termux-app#debugging

This build can't be used with Termux installed from F-Droid, right?

**Comment by machitgarha** [comment]
> Change log level to `Verbose` for Termux:API and then run api command and then run `adb logcat -d > logcat.txt`. **After that**, install Termux:API app from [here](https://github.com/termux/termux-api/actions/runs/5947876837) and then start app from launcher and then try running command again.
> 
> https://github.com/termux/termux-app#debugging

Thank you very much! It fixed the problem for me too!

**Comment by suikoy** [comment]
I have incurred in the same issue, is there a fix that I could apply? In my case I use termux-api just to show some popups with termux-toast.
I'm on Pixel 7 with Android 14 no root and Termux installed via F-Droid

**Comment by DobryninVyacheslav** [comment]
@suikoy, try this command: `apt update; apt install termux-api termux-am`
As written in the https://github.com/termux/termux-api/issues/638#issuecomment-1868383070
It helped me

**Comment by suikoy** [comment]
Yes it's works like a charme! Many thanks Dobrynin

**Comment by amirhosein-vedadi** [comment]
To address the issue of termux-api commands not working and hanging in the terminal without any visible output, I suggest the following solution:

1- Uninstall the termux app from your device.
2- Install the latest version of [termux-api ](https://github.com/termux/termux-api)from F-Droid.
3- After installing termux-api, install the latest version of the [termux app](https://github.com/termux/termux-app) from F-Droid.

I followed this procedure on my Redmi 10c device running Android 13 and was able to resolve the issue.

**Comment by WA1870** [comment]
From What I Know Termux-api Commands Hang Because Termux:APi App Is Not Running. I Have Same Problem Because First DuraSpeed Restricted Termux:API So I Uninstall DuraSpeed But Now Battery Manager Restricted Termux:API If I Disabled Battery Manager. So  I Have To Reboot Mobile To Relaunch Termux:API App But Now I Found A Solution For This. I Don't Have To Reboot Mobile. But Battery Manager Kill Termux:API But This Is Best Solution For Me.

https://youtu.be/mL4FvCVJ4Os?si=i3uDgmeUXS2ERfF7

**Comment by david-shu** [comment]
I encountered the same problem. Through the following methods:

-Install termux API in termux terminal:  pkg install termux-api
-Install termux-api APK under Android
-Run under termux terminal:  termux-api-start

Now the notification has been successfully popped up:

-Run under termux terminal:  termux-toast hello

**Comment by DintonBQ** [comment]
> To address the issue of termux-api commands not working and hanging in the terminal without any visible output, I suggest the following solution:
> 
> 1- Uninstall the termux app from your device. 2- Install the latest version of [termux-api ](https://github.com/termux/termux-api)from F-Droid. 3- After installing termux-api, install the latest version of the [termux app](https://github.com/termux/termux-app) from F-Droid.
> 
> I followed this procedure on my Redmi 10c device running Android 13 and was able to resolve the issue.

thanks, fixed

**Comment by mrkeuz** [comment]
In my case A13 (pure AOSP) duraspeed lock the app. And then I cannot  start it - got this/
```
~ $ termux-api-start
Starting service: Intent { cmp=com.termux.api/.KeepAliveService }
Error: Unable to launch app com.termux.api/10206 for service Intent { cmp=com.termux.api/.KeepAliveService }: process is bad
```

Also I cannot remove limit in duraspeed because Termux:API NOT regular app (don't know how this properly named). It does not show Termux:API in app list. 

Only help: 
- Disable duraspeed complete 
- Then run termux-api-start from termux

Reinstall does not help. Limitation in duraspeed cached and dit not reset. 

Maybe would be good change distribution and post Termux:API as regular app, with icon and app behavior - I don't know how correct this named (new in android). Thanks!
